// @microsoft/teams.apps/dist/plugins/http/plugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpPlugin = void 0;
const http_1 = __importDefault(require("http"));
const cors_1 = __importDefault(require("cors"));
const express_1 = __importDefault(require("express"));
const teams_api_1 = require("@microsoft/teams.api");
const $http = __importStar(require("@microsoft/teams.common/http"));
const package_json_1 = __importDefault(require("../../../package.json"));
const types_1 = require("../../types");
const stream_1 = require("./stream");
/**
 * Can send/receive activities via http
 */
let HttpPlugin = class HttpPlugin {
    logger;
    client;
    manifest;
    botToken;
    graphToken;
    $onError;
    $onActivity;
    get;
    post;
    patch;
    put;
    delete;
    route;
    use;
    get server() {
        return this._server;
    }
    _server;
    get port() {
        return this._port;
    }
    _port;
    express;
    pending = {};
    constructor(server) {
        this.express = (0, express_1.default)();
        this._server = server || http_1.default.createServer();
        this._server.on('request', this.express);
        this.get = this.express.get.bind(this.express);
        this.post = this.express.post.bind(this.express);
        this.patch = this.express.patch.bind(this.express);
        this.put = this.express.put.bind(this.express);
        this.delete = this.express.delete.bind(this.express);
        this.route = this.express.route.bind(this.express);
        this.use = this.express.use.bind(this.express);
        this.express.use((0, cors_1.default)());
        this.express.use('/api*', express_1.default.json());
        this.express.post('/api/messages', this.onRequest.bind(this));
    }
    /**
     * serve static files
     * @param path the url path to serve
     * @param dist the dist file path to serve
     */
    static(path, dist) {
        this.express.use(path, express_1.default.static(dist));
        return this;
    }
    async onStart({ port }) {
        this._port = port;
        this.express.get('/', (_, res) => {
            res.send(this.manifest);
        });
        return await new Promise((resolve, reject) => {
            this._server.on('error', (err) => {
                this.$onError({ error: err });
                reject(err);
            });
            this._server.listen(port, () => {
                this.logger.info(`listening on port ${port} 🚀`);
                resolve();
            });
        });
    }
    onStop() {
        this._server.close();
    }
    onError({ error, activity }) {
        if (!activity)
            return;
        const res = this.pending[activity.id];
        if (!res) {
            return;
        }
        if (!res.headersSent) {
            res.status(500).send(error.message);
        }
        delete this.pending[activity.id];
    }
    onActivityResponse({ response, activity }) {
        const res = this.pending[activity.id];
        if (!res) {
            return;
        }
        if (!res.headersSent) {
            res.status(response.status || 200).send(JSON.stringify(response.body || null));
        }
        delete this.pending[activity.id];
    }
    async send(activity, ref) {
        const api = new teams_api_1.Client(ref.serviceUrl, this.client.clone({
            token: this.botToken,
        }));
        activity = {
            ...activity,
            from: ref.bot,
            conversation: ref.conversation,
        };
        if (activity.id) {
            const res = await api.conversations
                .activities(ref.conversation.id)
                .update(activity.id, activity);
            return { ...activity, ...res };
        }
        const res = await api.conversations.activities(ref.conversation.id).create(activity);
        return { ...activity, ...res };
    }
    createStream(ref) {
        return new stream_1.HttpStream(new teams_api_1.Client(ref.serviceUrl, this.client.clone({
            token: this.botToken,
        })), ref, this.logger);
    }
    /**
     * validates an incoming http request
     * @param req the incoming http request
     * @param res the http response
     */
    async onRequest(req, res, _next) {
        const authorization = req.headers.authorization?.replace('Bearer ', '');
        if (!authorization && process.env.NODE_ENV !== 'local') {
            res.status(401).send('unauthorized');
            return;
        }
        const activity = req.body;
        const token = authorization
            ? new teams_api_1.JsonWebToken(authorization)
            : {
                appId: '',
                from: 'azure',
                fromId: '',
                serviceUrl: activity.serviceUrl || '',
                isExpired: () => false,
            };
        this.pending[activity.id] = res;
        this.$onActivity({
            sender: this,
            activity,
            token,
        });
    }
};
exports.HttpPlugin = HttpPlugin;
__decorate([
    (0, types_1.Logger)(),
    __metadata("design:type", Object)
], HttpPlugin.prototype, "logger", void 0);
__decorate([
    (0, types_1.Dependency)(),
    __metadata("design:type", $http.Client)
], HttpPlugin.prototype, "client", void 0);
__decorate([
    (0, types_1.Dependency)(),
    __metadata("design:type", Object)
], HttpPlugin.prototype, "manifest", void 0);
__decorate([
    (0, types_1.Dependency)({ optional: true }),
    __metadata("design:type", Function)
], HttpPlugin.prototype, "botToken", void 0);
__decorate([
    (0, types_1.Dependency)({ optional: true }),
    __metadata("design:type", Function)
], HttpPlugin.prototype, "graphToken", void 0);
__decorate([
    (0, types_1.Event)('error'),
    __metadata("design:type", Function)
], HttpPlugin.prototype, "$onError", void 0);
__decorate([
    (0, types_1.Event)('activity'),
    __metadata("design:type", Function)
], HttpPlugin.prototype, "$onActivity", void 0);
exports.HttpPlugin = HttpPlugin = __decorate([
    (0, types_1.Plugin)({
        name: 'http',
        version: package_json_1.default.version,
        description: 'the default plugin for sending/receiving activities',
    }),
    __metadata("design:paramtypes", [http_1.default.Server])
], HttpPlugin);
