// @microsoft/teams.apps/dist/router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Router = void 0;
const routes_1 = require("./routes");
class Router {
    routes = [];
    /**
     * select routes that match the inbound activity
     * @param activity the inbound activity
     */
    select(activity) {
        return this.routes
            .filter((r) => r.select(activity))
            .map((r) => r.callback);
    }
    /**
     * register a new route
     * @param route the route to register
     */
    register(route) {
        this.routes.push(route);
        return this;
    }
    /**
     * register a middleware
     * @param callback the callback to invoke
     */
    use(callback) {
        this.register({
            select: () => true,
            callback,
        });
        return this;
    }
    /**
     * register an activity route
     * @param event event to subscribe to
     * @param callback the callback to invoke
     */
    on(event, callback) {
        this.register({
            name: event,
            select: (activity) => {
                if (event === 'activity') {
                    return true;
                }
                if (event === activity.type) {
                    return true;
                }
                if (activity.type === 'conversationUpdate') {
                    return event === activity.channelData?.eventType;
                }
                if (activity.type === 'installationUpdate') {
                    return event === `install.${activity.action}`;
                }
                if (activity.type === 'messageDelete') {
                    return event === activity.channelData?.eventType;
                }
                if (activity.type === 'messageUpdate') {
                    return event === activity.channelData?.eventType;
                }
                if (activity.type === 'event') {
                    return event === routes_1.EVENT_ALIASES[activity.name];
                }
                if (activity.type === 'invoke') {
                    if (event === routes_1.INVOKE_ALIASES[activity.name]) {
                        return true;
                    }
                    if (activity.name === 'fileConsent/invoke') {
                        return event === `file.consent.${activity.value.action}`;
                    }
                    if (activity.name === 'composeExtension/submitAction') {
                        return event === `message.ext.${activity.value.botMessagePreviewAction}`;
                    }
                    if (activity.name === 'message/submitAction') {
                        return event === `message.submit.${activity.value.actionName}`;
                    }
                }
                // custom routes
                if (event === 'mention' && activity.entities?.some((e) => e.type === 'mention')) {
                    return (activity.entities?.find((e) => e.type === 'mention' && e.mentioned.id === activity.recipient.id) !== undefined);
                }
                return false;
            },
            callback,
        });
        return this;
    }
}
exports.Router = Router;
