// @microsoft/teams.common/dist/logging/ansi.js
'use strict';

var ANSI = /* @__PURE__ */ ((ANSI2) => {
  ANSI2["Reset"] = "\x1B[0m";
  ANSI2["Bold"] = "\x1B[1m";
  ANSI2["BoldReset"] = "\x1B[22m";
  ANSI2["Italic"] = "\x1B[3m";
  ANSI2["ItalicReset"] = "\x1B[23m";
  ANSI2["Underline"] = "\x1B[4m";
  ANSI2["UnderlineReset"] = "\x1B[24m";
  ANSI2["Strike"] = "\x1B[9m";
  ANSI2["StrikeReset"] = "\x1B[29m";
  ANSI2["ForegroundReset"] = "\x1B[0m";
  ANSI2["BackgroundReset"] = "\x1B[0m";
  ANSI2["ForegroundBlack"] = "\x1B[30m";
  ANSI2["BackgroundBlack"] = "\x1B[40m";
  ANSI2["ForegroundRed"] = "\x1B[31m";
  ANSI2["BackgroundRed"] = "\x1B[41m";
  ANSI2["ForegroundGreen"] = "\x1B[32m";
  ANSI2["BackgroundGreen"] = "\x1B[42m";
  ANSI2["ForegroundYellow"] = "\x1B[33m";
  ANSI2["BackgroundYellow"] = "\x1B[43m";
  ANSI2["ForegroundBlue"] = "\x1B[34m";
  ANSI2["BackgroundBlue"] = "\x1B[44m";
  ANSI2["ForegroundMagenta"] = "\x1B[35m";
  ANSI2["BackgroundMagenta"] = "\x1B[45m";
  ANSI2["ForegroundCyan"] = "\x1B[36m";
  ANSI2["BackgroundCyan"] = "\x1B[46m";
  ANSI2["ForegroundWhite"] = "\x1B[37m";
  ANSI2["BackgroundWhite"] = "\x1B[47m";
  ANSI2["ForegroundGray"] = "\x1B[90m";
  ANSI2["ForegroundDefault"] = "\x1B[39m";
  ANSI2["BackgroundDefault"] = "\x1B[49m";
  return ANSI2;
})(ANSI || {});

exports.ANSI = ANSI;
